### PR TITLE
Update PyQt to 5.15

### PR DESCRIPTION
**Pull request by andfoy**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by andfoy** [comment]
@conda-forge-admin, please re-render

**Comment by andfoy** [comment]
@conda-forge-admin, please re-render

File: recipe/charts_bld.bat
Comment (ccordoba12) [inline recipe/charts_bld.bat:1]: Any reason to change the name of the scripts used before? I mean, before we had (for instance) `bld-pyqtcharts.bat` and you changed that to `charts_bld.bat`, which makes reviewing your changes harder.

If it's not too much to ask, could you revert to the previous naming? That would also make easier to follow the git history.

File: recipe/meta.yaml
Comment (ccordoba12) [inline recipe/meta.yaml:39]: I think we should keep this skip so that the regro bot doesn't try to add support for that architecture here.

File: recipe/meta.yaml
Comment (ccordoba12) [inline recipe/meta.yaml:63]: Perhaps this should be renamed to `pyqt-main`? I mean, that way it'd be easier for users and developers to know that this is the package that goes with `qt-main`.

**Review by ccordoba12 — COMMENTED** [review]
Thanks for all your hard work on this @andfoy!

I left some minor questions/suggestions of form, otherwise looks good to me.

File: recipe/meta.yaml
Comment (andfoy) [reply recipe/meta.yaml:63]: The previous package was named `pyqt-impl`, I don't know if we should rename them

File: recipe/charts_bld.bat
Comment (andfoy) [reply recipe/charts_bld.bat:1]: The build system used to compile PyQt is completely different from the one used before, so the diff will be deleting and adding new lines. I can rename them, but for reviewing purposes, it won't make much difference

File: recipe/meta.yaml
Comment (ccordoba12) [reply recipe/meta.yaml:64]: @isuruf, thoughts about this? I think we should rename this package to make naming consistent.

Comment (isuruf) [reply recipe/meta.yaml:64]: I thought we wanted to go with upstreaming name which `pyqt = qt-main + python` and `pyqtwebengine = qt-webengine + pyqt`.

Comment (ccordoba12) [reply recipe/meta.yaml:64]: Right, I forgot about it. I'm ok with that too.

@andfoy?

Comment (andfoy) [reply recipe/meta.yaml:64]: I also forgot about that one, however, we should we do with respect to pyqt-charts? Create `pyqtcharts = pyqt + pyqt-charts`?

Comment (ccordoba12) [reply recipe/meta.yaml]: Then I'd say we should go with your suggestion above.

**Comment by ccordoba12** [comment]
One last question for @isuruf: how should packages that depend on `pyqtwebengine` handle this transition for them to be able to work with PyQt 5.12 and 5.15?

In other words: those packages will need to add the dependency on `pyqtwebengine` explicitly for them to work with both versions (right now they only need to depend on `pyqt`). But this could catch feedstock maintainers off guard if they have not followed this process. So how do you think we should let them about it?

**Comment by isuruf** [comment]
Good question. I guess there's no easy way to avoid it. Maybe we should add a `pyqtwebengine` empty meta package for 5.12 so that maintainers can opt in early.

File: recipe/meta.yaml
Comment (isuruf) [inline recipe/meta.yaml:40]: Nit. Can we remove the jinja here? It's hard to review otherwise

File: recipe/meta.yaml
Comment (isuruf) [inline recipe/meta.yaml:37]: Why this change in the selector?

**Comment by andfoy** [comment]
> Good question. I guess there's no easy way to avoid it. Maybe we should add a pyqtwebengine empty meta package for 5.12 so that maintainers can opt in early.

Should I send a separate PR updating the already existing recipe to include pyqtwebengine as an empty package?

**Comment by ccordoba12** [comment]
> I guess there's no easy way to avoid it.

Perhaps a blog post? And then tweet about it?

> Although, the pyqtwebengine package already exists: https://anaconda.org/conda-forge/pyqtwebengine

Yeah, there is. And it contains the bindings for `qtwebengine`. So maintainers will only need to add that dependency explicitly to their recipes (even if it doesn't add anything new for 5.12).

**Review by ccordoba12 — APPROVED** [review]
Thanks a lot @andfoy for your hard work on this!

**Comment by ccordoba12** [comment]
I'll let @isuruf give the final word and merge here (I have nothing else to add).

File: recipe/meta.yaml
Comment (isuruf) [inline recipe/meta.yaml:64]: Same here. Can you change all of these?

**Comment by Tobias-Fischer** [comment]
Many thanks for the hard work @andfoy! Looking forward to seeing this merged - and even more to the migration to 5.15 :)

File: recipe/bld-pyqt-sip.bat
Comment (h-vetinari) [inline recipe/bld-pyqt-sip.bat:2]: Is there a reason all the `.bat` files don't contain checks for the errorlevels anymore?

To fail early upon problems (rather than run into obscure problems later), I think this should be reinstated for every non-trivial call.
```suggestion
%PYTHON% -m pip install . -vv
if %ERRORLEVEL% neq 0 exit 1
```

File: recipe/bld-pyqt.bat
Comment (h-vetinari) [inline recipe/bld-pyqt.bat:11]: Any reason not to indent multiline commands?
```suggestion
call sip-build ^
    --verbose ^
    --confirm-license ^
    --disable QtNfc ^
    --no-make
if %ERRORLEVEL% neq 0 exit 1
```

File: recipe/bld-pyqt.bat
Comment (h-vetinari) [inline recipe/bld-pyqt.bat:15]:
```suggestion
jom
if %ERRORLEVEL% neq 0 exit 1
```

File: recipe/bld-pyqt.bat
Comment (h-vetinari) [inline recipe/bld-pyqt.bat:16]:
```suggestion
jom install
if %ERRORLEVEL% neq 0 exit 1
```

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:348]: could add
```
  feedstock-name: pyqt
```
(below maintainers) so that the REAME doesn't pick up "pyqt-split".

**Review by h-vetinari — COMMENTED** [review]
Some observations from taking a quick look. Great to see this being worked on!

File: recipe/yum_requirements.txt
Comment (h-vetinari) [inline recipe/yum_requirements.txt:8]: What's the benefit of breaking the previously alphabetical order here? A lot of deps just get shuffled around with no actual change?

Comment (h-vetinari) [reply recipe/yum_requirements.txt:8]: Aside from this nit (and not being the right person to review this): LGTM!

**Comment by andfoy** [comment]
@isuruf @ccordoba12 @h-vetinari Do you have any other comments?

**Comment by ccordoba12** [comment]
None from my side. Also, thanks @h-vetinari for your review! Nice to see the community taking part of this effort.

@isuruf, are we ready to merge?

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:11]:
```suggestion
{% set version = "5.15.4" %}
{% set sip_version = "12.9.0" %}
{% set webengine_version = "5.15.4" %}
{% set charts_version = "5.15.5" %}
```

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:19]:
```suggestion
  - url: https://pypi.io/packages/source/P/PyQt{{ version[0] }}/PyQt{{ version[0] }}-{{ version }}.tar.gz
```

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:23]:
```suggestion
  - url:  https://pypi.io/packages/source/P/PyQt5-sip/PyQt5_sip-{{ sip_version }}.tar.gz
```

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:27]:
```suggestion
  - url: https://pypi.io/packages/source/P/PyQtWebEngine/PyQtWebEngine-{{ webengine_version }}.tar.gz
```

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:31]:
```suggestion
  - url: https://pypi.io/packages/source/P/PyQtChart/PyQtChart-{{ charts_version }}.tar.gz
```

**Review by h-vetinari — COMMENTED** [review]
Just noticed one more thing following https://github.com/conda-forge/pyqt-feedstock/pull/107/commits/7b1b0607cb74748c222be6be680bc0a6ca0ddb0d - I'm not a big fan of putting the package name in variables. In almost all cases, this is a fake kind of flexibility - it's not like the PyPI URL (or the package-outputname) is something we can change just by setting a different value.

**Comment by andfoy** [comment]
@isuruf @ccordoba12 @h-vetinari do you have any other comments?

**Comment by h-vetinari** [comment]
> @isuruf @ccordoba12 @h-vetinari do you have any other comments?

I think we're mainly waiting for @isuruf to have a look (who is in high demand for reviews and much other work throughout cf, so it can take a while).

**Comment by isuruf** [comment]
We need to make an announcement that we'll be dropping `pyqtwebengine` and `pyqtcharts` from the main `pyqt` package to match upstream PyPI package and to add them as dependencies (regardless of the pyqt version) if a package needs them.

**Comment by isuruf** [comment]
Otherwise looks good to merge

**Comment by h-vetinari** [comment]
> We need to make an announcement

Blogpost on conda-forge.github.io, or where did you have in mind?

Would be cool if we could add that message to the 5.15 migrator itself, but I don't think that's implemented yet? (I'm only aware of the "special" [migrators](https://github.com/regro/cf-scripts/tree/master/conda_forge_tick/migrators) having custom pr bodies, not the [ones](https://github.com/conda-forge/conda-forge-pinning-feedstock/tree/main/recipe/migrations) in the pinning feedstock)

**Comment by isuruf** [comment]
Announcements page https://github.com/conda-forge/conda-forge.github.io/blob/main/src/user/announcements.rst

**Comment by isuruf** [comment]
Yes, a migrator for qt 5.15 would be good to have so that pyqt=5.15 can be used in conjunction with other packages.

**Comment by h-vetinari** [comment]
> Announcements page https://github.com/conda-forge/conda-forge.github.io/blob/main/src/user/announcements.rst

TIL... 🙃

@andfoy, probably'd be best if you prepare a PR to https://github.com/conda-forge/conda-forge.github.io for the announcement?

**Comment by andfoy** [comment]
Thanks for all your help reviewing all the packages throughout these months! It's great to see that the final step is getting merged!

> Announcements page 

Regarding the announcements page, should I leave today's date as the one on the announcement, or will the announcement be published once this is merged?

**Comment by h-vetinari** [comment]
> Regarding the announcements page, should I leave today's date as the one on the announcement

The exact date is probably not hyper-relevant, but I'd add e.g. tomorrows date and open a PR with that. The order of merging can then still be decided, but I guess the idea is to keep it close-to-concurrent.

**Comment by andfoy** [comment]
The announcement PR is now open: https://github.com/conda-forge/conda-forge.github.io/pull/1734

File: recipe/meta.yaml
Comment (hmaarrfk) [inline recipe/meta.yaml:126]: This is still somewhat problematic right now.

Yesterday, I thought I could build the meta package, `qt` that depended on both `qt-main` and `qt-webengine`. The problem is that 
* qt-main is version 5.15.2 
* qt-webengine is version 5.15.4

They both depend on the `qt` package with `x.x.x` as pin.

The issue is that qt `5.15.4` seems to be a commercial only release that doesn't have a download file available on the opensource download page. I'm not too sure how to relax this constraint. Perhaps we should relax the constraint to `x.x` and not `x.x.x`.

Comment (h-vetinari) [reply recipe/meta.yaml:126]: So both [qt-main](https://github.com/conda-forge/qt-main-feedstock/blame/main/recipe/meta.yaml#L51) and [qt-webengine](https://github.com/conda-forge/qt-webengine-feedstock/blame/main/recipe/meta.yaml#L31)  already use the `x.x` pinning, so it seems to me that this should be possible...?

However, the [run-constraint](https://github.com/conda-forge/qt-main-feedstock/blame/main/recipe/meta.yaml#L175) of qt-main -> qt was added by @isuruf precisely in the context of the [split](https://github.com/conda-forge/qt-main-feedstock/commit/aec1b1044b7d4552c8c4022e3e1177df74aa2a17) into `qt-main`, so not sure if I'm missing something there.

Comment (hmaarrfk) [reply recipe/meta.yaml:126]: I guess you are missing the timeline of events.

I raised this concern, and worked through some of the issues to address it. Still need to migrate qt-webengine to the new icu

**Comment by h-vetinari** [comment]
Alrighty, announcement has been merged. This should be good now? @conda-forge/pyqt @isuruf

**Review by duncanmmacleod — APPROVED** [review]

**Comment by duncanmmacleod** [comment]
@conda-forge/pyqt, can this be merged?

**Comment by hmaarrfk** [comment]
I think you can install 5.15.2 and icu69. Should be fine for many people.

**Comment by ccordoba12** [comment]
Sorry, I guess you're saying that we should pin this package to `qt-main` 5.15.2 while PR https://github.com/conda-forge/qt-webengine-feedstock/pull/6 is sorted out, right?

**Comment by hmaarrfk** [comment]
Yes that way it will be forward compatible

**Comment by ccordoba12** [comment]
Ok, could you look at commit ec85b81 to see if that's exactly what you have in mind?

File: recipe/meta.yaml
Comment (hmaarrfk) [inline recipe/meta.yaml:132]:
```suggestion
        - qt-main 5.15
```

You should be able to pin runtime a little looser

File: recipe/meta.yaml
Comment (hmaarrfk) [inline recipe/meta.yaml:238]:
```suggestion
        - qt-main 5.15
```

File: recipe/meta.yaml
Comment (hmaarrfk) [inline recipe/meta.yaml:323]:
```suggestion
        - qt-main 5.15
```

**Review by hmaarrfk — COMMENTED** [review]
i think you can pin runtime a little looser

**Comment by andfoy** [comment]
I think this should suffice for this case:

```
- {{ pin_compatible("qt-main", min_pin="x.x", max_pin="x") }}
```

**Comment by ccordoba12** [comment]
Ok, @andfoy suggestion didn't work because it's pulling qt-main 5.15.3. Trying @hmaarrfk's suggestion now.

**Comment by ccordoba12** [comment]
Unfortunately, that also pulls qt-main 5.15.3, so going back to my original pin, which was successful yesterday. Since a lot of people is waiting for this, I think we should merge with a hard pin on qt-main 5.15.2 for now.

@hmaarrfk, do you agree?

**Comment by hmaarrfk** [comment]
I guess the thing is that the icu migration is half way through https://conda-forge.org/status/#icu70 and I would like to be able to use this package in ICU69 and ICU70.

I guess our current understanding is:

| icu | qt-main | qt-webenigne |
|------|------|-------| 
| 69 | 5.15.2   | 5.15.4 build 0 or 1 | 
| 70 | 5.15.3   | 5.15.4 build 2  -- was pulled for OSX | 

But in truth, qt-webengine was never built with system icu support. We are working through this at the moment..

So you have 3 options: 
* If you compile with qt-main `5.15.2`, pin to `5.15.*` then this system will be compatible with ICU69 and ICU70. In the test time, it will pull in icu70. That is OK.
* If you compile with qt-main `5.15.3`, pin to `5.15.*` then this system will only be compatible with ICU70.
* If you compile with qt-main `5.15.2` and pin to `5.15.2`, then you will only be compatible with icu69.

Therefore the most compatible option, is in my mind, to build WITH `5.15.2` and pin to `5.15`

**Comment by hmaarrfk** [comment]
Thank you for helping so much!

**Comment by ccordoba12** [comment]
Ok, everything is green now, so I'll merge tomorrow unless I hear any objections about it.

**Comment by ccordoba12** [comment]
Ok, the time to merge has come!

Thanks @andfoy for all your hard work on this!! This is a terrific improvement for all projects that use PyQt in conda-forge.

Thanks also to @isuruf, @hmaarrfk, @h-vetinari and all others that reviewed this PR and helped @andfoy to finish it!

**Comment by andfoy** [comment]
Thanks everyone for your help!

**Comment by SrNetoChan** [comment]
Great work @andfoy and others. Thanks!

**Pull request merged**
